app/main.py

```diff
@@ -130,85 +133,42 @@ def battle(knightsConfig):
             arthur["protection"] += arthur["potion"]["effect"]["protection"]
 
-        if "hp" in arthur["potion"]["effect"]:
-            arthur["hp"] += arthur["potion"]["effect"]["hp"]
-
-    # mordred
-    mordred = knightsConfig["mordred"]
-
-    # apply armour
-    mordred["protection"] = 0
-    for a in mordred["armour"]:
-        mordred["protection"] += a["protection"]
-
-    # apply weapon
-    mordred["power"] += mordred["weapon"]["power"]
-
-    # apply potion if exist
-    if mordred["potion"] is not None:
-        if "power" in mordred["potion"]["effect"]:
-            mordred["power"] += mordred["potion"]["effect"]["power"]
-
-        if "protection" in mordred["potion"]["effect"]:
-            mordred["protection"] += mordred["potion"]["effect"]["protection"]
-
-        if "hp" in mordred["potion"]["effect"]:
-            mordred["hp"] += mordred["potion"]["effect"]["hp"]
 
-    # red_knight
-    red_knight = knightsConfig["red_knight"]
 
-    # apply armour
-    red_knight["protection"] = 0
-    for a in red_knight["armour"]:
-        red_knight["protection"] += a["protection"]
-
-    # apply weapon
-    red_knight["power"] += red_knight["weapon"]["power"]
-
-    # apply potion if exist
-    if red_knight["potion"] is not None:
-        if "power" in red_knight["potion"]["effect"]:
-            red_knight["power"] += red_knight["potion"]["effect"]["power"]
+def create_knights(attributes: dict) -> list:
+    knights = []
+    for attr in attributes.values():
+        knight = Knight(name=attr["name"],
+                        power=attr["power"],
+                        hp=attr["hp"],
+                        armour=attr["armour"],
+                        weapon=attr["weapon"],
+                        potion=attr["potion"])
+        knight.preparation()
+        knights.append(knight)
+    return knights
 
-        if "protection" in red_knight["potion"]["effect"]:
-            red_knight["protection"] += red_knight["potion"]["effect"]["protection"]
 
-        if "hp" in red_knight["potion"]["effect"]:
-            red_knight["hp"] += red_knight["potion"]["effect"]["hp"]
+def check_hp(knight: Knight) -> None:
+    if knight.hp <= 0:
+        knight.hp = 0
 
-    # -------------------------------------------------------------------------------
-    # BATTLE:
 
-    # 1 Lancelot vs Mordred:
-    lancelot["hp"] -= mordred["power"] - lancelot["protection"]
-    mordred["hp"] -= lancelot["power"] - mordred["protection"]
+def fight(knight1: Knight, knight2: Knight) -> None:
+    knight1.hp -= knight2.power - knight1.protection
+    knight2.hp -= knight1.power - knight2.protection
+    check_hp(knight1)
+    check_hp(knight2)
 
-    # check if someone fell in battle
-    if lancelot["hp"] <= 0:
-        lancelot["hp"] = 0
 
-    if mordred["hp"] <= 0:
-        mordred["hp"] = 0
+def battle(knights: dict) -> dict:
+    knights = create_knights(knights)
+    fight(knights[0], knights[2])
+    fight(knights[1], knights[3])
```

Comment:
Due to task fighting must be based on name, not indexes. If I change order of elements i will get wrong output